### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by LeonIllya**

### CODE REVIEW

File: src/main/java/book/store/model/Order.java [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+@EqualsAndHashCode(of = {"id"})
+@Entity
+@SQLDelete(sql = "UPDATE orders SET is_deleted = TRUE WHERE id=?")
+@SQLRestriction(value = "is_deleted = FALSE")
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java:35]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
    @JoinColumn(name = "user_id", nullable = false)
```

File: src/main/java/book/store/model/Order.java [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+    @JoinColumn(nullable = false)
+    @MapsId
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java:37]:
```suggestion
```
`@MapsId` annotation in Hibernate is used to obtain a one-to-one relationship between two entities by mapping the primary key of one entity to the foreign key of another entity.  This annotation is used when we have to use a shared primary key between two entities.

File: src/main/java/book/store/model/Order.java [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    private User user;
+    @Column(nullable = false)
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java:39]:
```suggestion
    @Column(nullable = false)
    @Enumerated(EnumType.STRING)
    private Status status = Status.PENDING;
```

File: src/main/java/book/store/model/Order.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    private Status status;
+    @Column(nullable = false)
+    private BigDecimal total;
+    @Column(nullable = false)
+    private LocalDateTime orderDate;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/Order.java:44]:
```suggestion
    private LocalDateTime orderDate = LocalDateTime.now();
```
You can initialise orderDate here

File: src/main/java/book/store/model/OrderItem.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+@SQLRestriction(value = "is_deleted = false")
+@Table(name = "order_items")
+public class OrderItem {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "order_id", nullable = false)
+    private Order order;
+    @ManyToOne
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/model/OrderItem.java:33]:
```suggestion
    @ManyToOne(fetch = FetchType.LAZY, optional = false)
```

File: src/main/java/book/store/controller/OrderController.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Orders management", description = "Endpoint for managing orders")
+@RequiredArgsConstructor
+@RestController
+@Validated
+@RequestMapping("/orders")
+public class OrderController {
+    private final OrderService orderService;
+    private final OrderItemService orderItemService;
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/OrderController.java:32]: You actually don't need OrderItemService, you can implement all logic in OrderService

File: src/main/java/book/store/controller/OrderController.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    public OrderDto updateStatusOrder(@RequestBody @Valid UpdateStatusOrderDto requestDto,
+                                        @PathVariable @Positive Long orderId) {
+        return orderService.updateStatusOrder(orderId, requestDto);
+    }
+
+    @GetMapping("{orderId}/items")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public OrderItemDto getOrderItem(@PathVariable @Positive Long orderItemId,
+                                        @PathVariable @Positive Long orderId) {
+        return orderItemService.getOrderItem(orderItemId, orderId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/OrderController.java:60]:
```suggestion
        return orderService.findOrderItemsByOrder(user.getId(), orderId);
```

File: src/main/java/book/store/controller/OrderController.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    }
+
+    @GetMapping("{orderId}/items/{itemId}")
+    @PreAuthorize("hasRole('ROLE_USER')")
+    public Set<OrderItemDto> getAllOrderItemsByOrderId(@PathVariable @Positive Long orderId) {
+        return orderItemService.getAllOrderItemsByOrderId(orderId);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/controller/OrderController.java:66]:
```suggestion
        return orderService.findOrderItemById(orderId, itemId, user.getId());
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution!

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+
+        orderRepository.save(order);
+        return order;
+    }
+
+    private Set<OrderItem> createOrderItems(ShoppingCart shoppingCart, Order order) {
+        Set<OrderItem> orderItems = new HashSet<>();
+        for (CartItem cartItem : shoppingCart.getCartItems()) {
+            OrderItem orderItem = new OrderItem();
+            orderItem.setOrder(order);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java:141]: Could it be done by Mapper?

File: src/main/java/book/store/service/impl/OrderServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,151 @@
...
+                .filter(item -> item.getId().equals(itemId))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                    "Can`t find an order item by id: " + itemId));
+    }
+
+    private Order createOrder(User user, BigDecimal total,
+                                String shippingAddress, ShoppingCart shoppingCart) {
+        Order order = new Order();
+        order.setUser(user);
```
Comment (Elena-Bruyako) [inline src/main/java/book/store/service/impl/OrderServiceImpl.java:124]: Could it be done by Mapper?

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Let's consider to make some changes.

**Review by sarakhmen — APPROVED** [review]
lgtm!

**Pull request merged**